pkg/blockio/blockio.go

```diff
@@ -0,0 +1,456 @@
+	// If modified, check how to parse device node from expanded paths.
+	sysfsBlockDeviceIOSchedulerPaths = "/sys/block/*/queue/scheduler"
+)
+
+// tBlockDeviceInfo holds information on a block device to be configured.
+// As users can specify block devices using wildcards ("/dev/disk/by-id/*SSD*")
+// tBlockDeviceInfo.Origin is maintained for traceability: why this
+// block device is included in configuration.
+// tBlockDeviceInfo.DevNode contains resolved device node, like "/dev/sda".
+type tBlockDeviceInfo struct {
+	Major   int64
+	Minor   int64
+	DevNode string
+	Origin  string
+}
+
+// Our logger instance.
+var log grclog.Logger = grclog.NewLoggerWrapper(stdlog.New(os.Stderr, "[ blockio ] ", 0))
+
+// classBlockIO connects user-defined block I/O classes to
+// corresponding cgroups blockio controller parameters.
+var classBlockIO = map[string]cgroups.BlockIOParameters{}
+
+// SetLogger sets the logger instance to be used by the package.
+// Examples:
+//   // Log to standard logger:
+//   stdlog := log.New(os.Stderr, "blockio:", 0)
+//   blockio.SetLogger(goresctrllog.NewLoggerWrapper(stdlog))
+//   // Log to logrus:
+//   blockio.SetLogger(logrus.New())
+func SetLogger(l grclog.Logger) {
+	log = l
+}
+
+// SetConfigFromFile reads and applies blockio configuration from the
+// filesystem.
+func SetConfigFromFile(filename string, force bool) error {
+	if data, err := ioutil.ReadFile(filename); err == nil {
+		if err = SetConfigFromData(data, force); err != nil {
+			return fmt.Errorf("failed to set configuration from file %q: %s", filename, err)
+		}
+		return nil
+	} else {
+		return fmt.Errorf("failed to read config file %q: %v", filename, err)
+	}
+}
+
+// SetConfigFromData parses and applies configuration from data.
+func SetConfigFromData(data []byte, force bool) error {
+	config := &Config{}
+	if err := yaml.Unmarshal(data, &config); err != nil {
+		return err
+	}
+	return SetConfig(config, force)
+}
+
+// SetConfig scans available block devices and applies new configuration.
+func SetConfig(opt *Config, force bool) error {
+	if opt == nil {
+		// Setting nil configuration clears current configuration.
+		// SetConfigFromData([]byte(""), dontcare) arrives here.
+		classBlockIO = map[string]cgroups.BlockIOParameters{}
+		return nil
+	}
+
+	currentIOSchedulers, ioSchedulerDetectionError := getCurrentIOSchedulers()
+	if ioSchedulerDetectionError != nil {
+		log.Warnf("configuration validation partly disabled due to I/O scheduler detection error %#v", ioSchedulerDetectionError.Error())
+	}
+
+	classBlockIO = map[string]cgroups.BlockIOParameters{}
+	// Create cgroup blockio parameters for each blockio class
+	for class := range opt.Classes {
+		cgBlockIO, err := devicesParametersToCgBlockIO(opt.Classes[class], currentIOSchedulers)
+		if err != nil {
+			if force {
+				log.Warnf("ignoring: %v", err)
+			} else {
+				return err
+			}
+		}
+		classBlockIO[class] = cgBlockIO
+	}
+	return nil
+}
+
+// GetClasses returns block I/O class names
```

Comment:
nit: spotted one more missing punctuation from the public API documentation. This is the last nit like this, I promise 😊

Reply:
Thanks! Fixed. I should have went through all cases already in the previous patch. Now I've went through all the docstrings... :)